core/trino-spi/src/main/java/io/trino/spi/protocol/SpoolingManager.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.spi.protocol;
+
+import io.airlift.slice.Slice;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.URI;
+import java.util.Map;
+import java.util.Optional;
+
+public interface SpoolingManager
+{
+    SpooledSegmentHandle create(SpoolingContext context);
+
+    OutputStream createOutputStream(SpooledSegmentHandle handle)
+            throws Exception;
+
+    InputStream openInputStream(SpooledSegmentHandle handle)
+            throws IOException;
+
+    default void acknowledge(SpooledSegmentHandle handle)
+            throws IOException
+    {
+    }
+
+    default Optional<URI> directLocation(SpooledSegmentHandle handle)
+    {
+        return Optional.empty();
+    }
+
+    default Slice serialize(SpooledSegmentHandle handle)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    default SpooledSegmentHandle deserialize(Slice slice)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    default Map<String, String> headers(SpooledSegmentHandle deserialize)
```

Comment:
I see why you made this separate from `directLocation()`. Let's move it directly afterwards, since they logically go together.